### PR TITLE
Nav2 recovery server

### PR DESCRIPTION
**Pull request by SteveMacenski**
Implements a plugin-based recovery server and the following plugins
- spin
- backup

that still derive from the base action server of the recovery class. I think that this isn't the best way to go into the future, but baby steps to get to plugins here. 

Up for discussion for continued work:
- should each recovery broadcast its own action server independent of the recovery server to use? Optionally we can have 1 recovery action server and it indicates which plugin to use. Optionally we can support both, however add a Plugin definition method for `hasOwnServer()` to know between them.
- If we want them to all expose their own servers, how should we (or should we?) homologate them? I commented out the `executeRecovery()` method in the interface *for now* while we discuss 

~Right now this has 1 problem on my computer, but I'm convinced its a local issue that would work in someone elses environment but I'll validate before merge~ fixed

### CODE REVIEW

**Comment by SteveMacenski** [comment]
I know "it builds on my machine" is the reason we have CI, but CI's busted with issues relating to nav2_dynamic_params (now gone) and bt library (not touched) so I'm going to go ahead and just ignore it if that's ok with everyone else...

**Comment by mhpanah** [comment]
I am okay with either option as long as different recovery actions can be selected from the Behavior Tree. BT designer should be able to decide how to construct the tree and which recovery actions should be selected under different scenarios while the stack is running. 

Also, recovery actions should not just be limited to spin, backup, clear costmap, etc. A different planner by itself can be a recovery action.  For example: if the dwb local planner fails, the BT may decide to fallback to a different local planner rather than just spin or backup.

**Comment by SteveMacenski** [comment]
I dont think of multiple planners as recoveries. I think of them as fallback behaviors. That’s in scope for the multiple plugin loading on the controller/planner servers. 

Anyhow there’s technically nothing stopping you from doing any of those things. I would though challenge you to write a few more recovery plugins if you think you have some in mind beyond spin/backup that are generic enough for use. The recovery API was never very well used.

**Comment by SteveMacenski** [comment]
A couple examples of recovery plugins I'd thought of:
- Wait: just straight up just wait N time to see if whatever the issue is resolves itself
- Wait Until Stop: if recovery triggered due to person/moving thing in the way, look at the costmap and calculate some "entropy" heuristic and when it dies down enough allow to continue

but in general I have a really challenging time coming up with useful recoveries in this framework, or ones that generalize across many usecases

**Comment by SteveMacenski** [comment]
@mkhansen-intel @crdelsey pinging for review

File: nav2_bringup/launch/nav2_bringup_launch.py [hunk 1]
```diff
@@ -93,6 +93,7 @@ def generate_launch_description():
                                     'amcl',
                                     'controller_server',
                                     'planner_server',
+                                    'recoveries_server_node',
```
Comment (orduno) [inline nav2_bringup/launch/nav2_bringup_launch.py]: Can we remove `_node`?

File: nav2_core/include/nav2_core/recovery.hpp [hunk 2]
```diff
@@ -30,6 +32,8 @@ namespace nav2_core
 class Recovery
 {
 public:
+  using Ptr = std::shared_ptr<Recovery>;
```
Comment (orduno) [inline nav2_core/include/nav2_core/recovery.hpp]: `SharedPtr` instead?

File: nav2_core/include/nav2_core/recovery.hpp [hunk 3]
```diff
@@ -42,8 +46,8 @@ class Recovery
    * @param  costmap_ros A pointer to the costmap
    */
   virtual void configure(
-    const nav2_util::LifecycleNode * parent,
-    const std::string & name, tf2_ros::Buffer * tf) = 0;
+    const nav2_util::LifecycleNode::SharedPtr parent,
```
Comment (orduno) [inline nav2_core/include/nav2_core/recovery.hpp]: I think we've been using `rclcpp_lifecycle::LifecycleNode::SharedPtr`

File: nav2_core/include/nav2_core/recovery.hpp [hunk 4]
```diff
@@ -65,7 +69,7 @@ class Recovery
    * @param  name The name of this planner
    * @return true if successful, false is failed to execute fully
    */
-  virtual bool executeRecovery() = 0;
+  // virtual bool executeRecovery() = 0;
```
Comment (orduno) [inline nav2_core/include/nav2_core/recovery.hpp]: Maybe add a `TODO` on why is commented.

File: nav2_util/include/nav2_util/simple_action_server.hpp [hunk 5]
```diff
@@ -25,14 +25,14 @@
 namespace nav2_util
 {
 
-template<typename ActionT>
+template<typename ActionT, typename nodeT = rclcpp::Node>
```
Comment (orduno) [inline nav2_util/include/nav2_util/simple_action_server.hpp]: `typename nodeT = rclcpp::Node::SharedPtr`

File: nav2_util/include/nav2_util/simple_action_server.hpp [hunk 6]
```diff
@@ -25,14 +25,14 @@
...
 class SimpleActionServer
 {
 public:
   typedef std::function<void ()> ExecuteCallback;
 
   explicit SimpleActionServer(
-    rclcpp::Node::SharedPtr node,
+    typename nodeT::SharedPtr node,
```
Comment (orduno) [inline nav2_util/include/nav2_util/simple_action_server.hpp]: Oh, I see you define the pointer here. this is better.

File: nav2_recoveries/src/recovery_server.cpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+#include <memory>
+#include <string>
+#include <vector>
+#include "nav2_recoveries/recovery_server.hpp"
+
+namespace recovery_server
+{
+
+RecoveryServer::RecoveryServer()
+: nav2_util::LifecycleNode("nav2_recoveries", "", true),
```
Comment (orduno) [inline nav2_recoveries/src/recovery_server.cpp]: `: nav2_util::LifecycleNode("nav2_recoveries", "", false)` if we don't need the additional node.

File: nav2_bringup/launch/nav2_bringup_launch.py (hunk 1 above)
Comment (SteveMacenski) [reply nav2_bringup/launch/nav2_bringup_launch.py]: Yeah I can do that

File: nav2_core/include/nav2_core/recovery.hpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_core/include/nav2_core/recovery.hpp]: So this is `Ptr` to be consistent across the other adaptors. Local/global/goal checker use `Ptr`

File: nav2_core/include/nav2_core/recovery.hpp (hunk 3 above)
Comment (SteveMacenski) [reply nav2_core/include/nav2_core/recovery.hpp]: Can do, didn't notice that, wasn't intentional

**Comment by SteveMacenski** [comment]
For some reason won't let me commend on your last point on the constructor node thing - I did that to be consistent with the other 2 plugin factory servers. Do either of them need it too?

File: nav2_core/include/nav2_core/recovery.hpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_core/include/nav2_core/recovery.hpp]: So I was about to just change them all, but is also appears all the DWB plugins and BT  also use ::Ptr to describe shared ptrs, the only place defined in the stack that we do SharedPtr is in the lifecycle node wrapper (and subsequent rclcpp calls) which is a little too much to casually add to this PR. Can we make this out of scope?

File: nav2_recoveries/src/recovery_server.cpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+  plugin_loader_("nav2_core", "nav2_core::Recovery")
```
Comment (orduno) [inline nav2_recoveries/src/recovery_server.cpp]: Why does the loader needs to be a class member and not a local variable in `loadRecoveryPlugins`?

File: nav2_recoveries/include/nav2_recoveries/recovery_server.hpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  nav2_util::CallbackReturn on_deactivate(const rclcpp_lifecycle::State & state) override;
+  nav2_util::CallbackReturn on_cleanup(const rclcpp_lifecycle::State & state) override;
+  nav2_util::CallbackReturn on_shutdown(const rclcpp_lifecycle::State & state) override;
+  nav2_util::CallbackReturn on_error(const rclcpp_lifecycle::State & state) override;
+
+  std::shared_ptr<tf2_ros::Buffer> tf_;
+  std::shared_ptr<tf2_ros::TransformListener> transform_listener_;
+
+  // Plugins
+  std::vector<nav2_core::Recovery::Ptr> recoveries_;
```
Comment (orduno) [inline nav2_recoveries/include/nav2_recoveries/recovery_server.hpp]: The `recoveries_` could be `unique_ptr`s.

File: nav2_recoveries/src/recovery_server.cpp [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    }
+  }
+}
+
+nav2_util::CallbackReturn
+RecoveryServer::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  std::vector<nav2_core::Recovery::Ptr>::iterator iter;
```
Comment (orduno) [inline nav2_recoveries/src/recovery_server.cpp]: This should be equivalent
```cpp
for (auto & recovery : recoveries_) {
    recovery->activate();
}

```

File: nav2_recoveries/src/recovery_server.cpp [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    get_node_timers_interface());
+  tf_->setCreateTimerInterface(timer_interface);
+  transform_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_);
+
+  declare_parameter("costmap_topic",
+    rclcpp::ParameterValue(std::string("local_costmap/costmap_raw")));
+  declare_parameter("footprint_topic",
+    rclcpp::ParameterValue(std::string("local_costmap/published_footprint")));
+
+  std::vector<std::string> plugin_names{std::string("Spin"),
```
Comment (orduno) [inline nav2_recoveries/src/recovery_server.cpp]: Is there a reason why you want to assign the recovery a name here instead of being hardcoded on the recovery?

File: nav2_recoveries/src/recovery_server.cpp (hunk 10 above)
Comment (SteveMacenski) [reply nav2_recoveries/src/recovery_server.cpp]: I think I've mentioned this before, but I really don't like the use of `auto` nor the pythonic C++11 loops

File: nav2_recoveries/src/recovery_server.cpp (hunk 11 above)
Comment (SteveMacenski) [reply nav2_recoveries/src/recovery_server.cpp]: Because there are cases where you want multiple plugins of the same type and they need to be distinguishable

File: nav2_recoveries/plugins/back_up.cpp [hunk 12]
```diff
@@ -16,25 +16,29 @@
...
 
 using namespace std::chrono_literals;
 
 namespace nav2_recoveries
 {
 
-BackUp::BackUp(rclcpp::Node::SharedPtr & node, std::shared_ptr<tf2_ros::Buffer> tf)
-: Recovery<BackUpAction>(node, "BackUp", tf)
+BackUp::BackUp()
+: Recovery<BackUpAction>()
```
Comment (orduno) [inline nav2_recoveries/plugins/back_up.cpp]: simulate_ahead_time{2.0}

File: nav2_recoveries/src/recovery_server.cpp (hunk 8 above)
Comment (SteveMacenski) [reply nav2_recoveries/src/recovery_server.cpp]: To find the plugins, it needs to crawl the filesystem looking for different things so on construction limits us to doing this at most once and crashes immediately if certain inputs are invalid. Additionally I see a time in the not so distant future when we circle back to dynamic reconfigure like behavior that we can to load and unload plugins for tuning and having the member will allow for that.

But a more `this is why I did it` vs a justification reason is that's how I've always done it and that's how all the examples of pluginlib I've seen do it, so I've just never argued with it.

File: nav2_recoveries/plugins/back_up.cpp (hunk 12 above)
Comment (orduno) [reply nav2_recoveries/plugins/back_up.cpp]: Never mind, I saw spin has a bunch of variables as well, so just declare on the body for consistency.

Comment (SteveMacenski) [reply nav2_recoveries/plugins/back_up.cpp]: its set in the constructor https://github.com/ros-planning/navigation2/blob/8b8985ec6165606f75b5edb496db95ad67c0527a/nav2_recoveries/plugins/back_up.cpp#L29

File: nav2_recoveries/include/nav2_recoveries/recovery_server.hpp (hunk 9 above)
Comment (SteveMacenski) [reply nav2_recoveries/include/nav2_recoveries/recovery_server.hpp]: Yes, they could be, do you want this change?

**Comment by orduno** [comment]
>should each recovery broadcast its own action server independent of the recovery server to use? Optionally we can have 1 recovery action server and it indicates which plugin to use. Optionally we can support both, however add a Plugin definition method for hasOwnServer() to know between 
them.

If the recovery server can only execute one recovery at a time, I'd recommend keeping the action server on the recovery server. -- and avoid adding one more layer of complexity trying to synchronize/block the actions servers.

**Comment by SteveMacenski** [comment]
The main reason I find against having 1 recovery action exposed that gives the commands to the appropriate recovery behavior is trying to make a `recovery.msg` file to contain all the "stuff" someone might want for their recovery. Spin wants an angle, backup wants a distance, following XYZ wants XYZ, the message requirements would be limitless. Having each recovery expose its own action server pushes the ROS message types to the BT and the recovery plugin implementations. 

I dont think we need to worry greatly about async, if someone wants to call multiple at once, that's on them. The BT shouldn't allow for it and if they've found a way to do so, they know they're doing something strange and may expect blowback

File: nav2_core/include/nav2_core/recovery.hpp (hunk 2 above)
Comment (orduno) [reply nav2_core/include/nav2_core/recovery.hpp]: Yes, we can leave that for a later PR.

File: nav2_recoveries/src/recovery_server.cpp (hunk 8 above)
Comment (orduno) [reply nav2_recoveries/src/recovery_server.cpp]: Ok, mainly curious. I'm ok leaving it as a member variable.

File: nav2_recoveries/include/nav2_recoveries/recovery_server.hpp (hunk 9 above)
Comment (orduno) [reply nav2_recoveries/include/nav2_recoveries/recovery_server.hpp]: Yeah, I think we should use a unique ptr in this case
```cpp
std::vector<std::unique_ptr<nav2_core::Recovery>> recoveries_;
```

File: nav2_recoveries/src/recovery_server.cpp (hunk 10 above)
Comment (orduno) [reply nav2_recoveries/src/recovery_server.cpp]: :cry: whay is that? Those are two of the features I like the most from modern c++.

I don't have a strong objection if you really want to keep the iterator version.

**Comment by orduno** [comment]
Agree we don't have to define that on this PR.

Just occurred to me we could have an action server that serves different types of actions. This could be on the recovery server and each recovery registers its action definition and the callback. Also, we define if these can run in parallel or not.

**Comment by SteveMacenski** [comment]
That sounds... really painful or impossible with pluginlib, but I would be more than supportive of the attempt

File: nav2_recoveries/src/recovery_server.cpp [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    std::string("BackUp")};
+  std::vector<std::string> plugin_types{std::string("nav2_recoveries/Spin"),
```
Comment (orduno) [inline nav2_recoveries/src/recovery_server.cpp]: If we had a long list of plugins, I'd recommend setting a hash table instead of two separte vectors:

```cpp
std::unordered_map<std::string, std::string> plugins_{
    {"Spin", "nav2_recoveries/Spin"},
    ...}
))
```

Comment (SteveMacenski) [reply nav2_recoveries/src/recovery_server.cpp]: agreed, but not implemented in parameters yet in rclcpp

Comment (SteveMacenski) [reply nav2_recoveries/src/recovery_server.cpp]: I took that from the costmap layer method when I was surprised to find that the map parameter readin wasn't implemented in ROS2

Comment (SteveMacenski) [reply nav2_recoveries/src/recovery_server.cpp]: Once that's around I'll update it everywhere because you're right, this is ugly

File: nav2_recoveries/include/nav2_recoveries/recovery_server.hpp (hunk 9 above)
Comment (SteveMacenski) [reply nav2_recoveries/include/nav2_recoveries/recovery_server.hpp]: Will do

**Comment by SteveMacenski** [comment]
Travis is failing on downloads, and the `release_test` fails periodically on random things, never the same twice, and almost never related to this.

File: nav2_recoveries/plugins/back_up.cpp [hunk 14]
```diff
@@ -16,25 +16,29 @@
...
-  node_->get_parameter("simulate_ahead_time", simulate_ahead_time_);
 }
 
 BackUp::~BackUp()
 {
 }
 
+void BackUp::onConfigure()
+{
+  node_->declare_parameter("simulate_ahead_time");
```
Comment (bpwilcox) [inline nav2_recoveries/plugins/back_up.cpp]: Why don't we set the default here in the `declare_parameter` instead of in the constructor?

File: nav2_recoveries/src/recovery_server.cpp [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    std::string("nav2_recoveries/BackUp")};
```
Comment (bpwilcox) [inline nav2_recoveries/src/recovery_server.cpp]: Should this be `nav2_recoveries::Backup`?

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp [hunk 16]
```diff
@@ -110,12 +106,14 @@ class Recovery
...
-      *costmap_sub_, *footprint_sub_, tf_, node_->get_name(), "odom");
+      *costmap_sub_, *footprint_sub_, *tf_, node_->get_name(), "odom");
 
     vel_pub_ = node_->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 1);
+
+    onConfigure();
   }
 
-  void cleanup()
+  void cleanup() override
```
Comment (bpwilcox) [inline nav2_recoveries/include/nav2_recoveries/recovery.hpp]: Similar to `onConfigure` shouldn't there be an `onCleanup` to handle and de-allocate any members of the derived classes?

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp [hunk 17]
```diff
@@ -78,21 +73,22 @@ class Recovery
...
-  void configure()
+  void configure(
+    const rclcpp_lifecycle::LifecycleNode::SharedPtr parent,
+    const std::string & name, std::shared_ptr<tf2_ros::Buffer> tf) override
   {
-    RCLCPP_INFO(node_->get_logger(), "Configuring %s", recovery_name_.c_str());
+    RCLCPP_INFO(parent->get_logger(), "Configuring %s", name.c_str());
+
+    node_ = parent;
+    tf_ = tf;
```
Comment (bpwilcox) [inline nav2_recoveries/include/nav2_recoveries/recovery.hpp]: Cleanup the `node_` and `tf_` in the `cleanup()` call?

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp [hunk 18]
```diff
@@ -43,26 +44,20 @@ enum class Status : int8_t
...
-  : node_(node),
-    recovery_name_(recovery_name),
-    tf_(*tf),
-    action_server_(nullptr),
-    cycle_frequency_(10)
+  using ActionServer = nav2_util::SimpleActionServer<ActionT, rclcpp_lifecycle::LifecycleNode>;
+
+  Recovery()
+  : action_server_(nullptr),
+    cycle_frequency_(10),
```
Comment (bpwilcox) [inline nav2_recoveries/include/nav2_recoveries/recovery.hpp]: Seems like this should be made a parameter.

File: nav2_recoveries/plugins/back_up.cpp (hunk 14 above)
Comment (SteveMacenski) [reply nav2_recoveries/plugins/back_up.cpp]: No particular reason, I can update that

File: nav2_recoveries/src/recovery_server.cpp (hunk 15 above)
Comment (SteveMacenski) [reply nav2_recoveries/src/recovery_server.cpp]: No, I have them registered as `/`

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp (hunk 17 above)
Comment (SteveMacenski) [reply nav2_recoveries/include/nav2_recoveries/recovery.hpp]: node and tf are from the recovery server, which within that reset on cleanup. It wouldnt make sense for each of `N` recoveries to  reset the TF buffer or node ptr on shutdown

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp (hunk 18 above)
Comment (SteveMacenski) [reply nav2_recoveries/include/nav2_recoveries/recovery.hpp]: I agree, I actually have a ticket open to make this configurable per-recovery, but hasn't been a priority recently

**Comment by SteveMacenski** [comment]
@bpwilcox the abstract interface from `nav2_core` is the _real_ recovery interface. The `Recovery.hpp` wrapper around it wraps the action server, but that's not the _real_ base interface to use, that's another level on top to reduce code duplication if that recovery is an action server, but there's nothing requiring that to be true. It could be a topic, service, action, etc. 

A plugin designer could choose to use the `Recovery.hpp` wrapped interface if it makes sense, but there may be quick things that should be services, where the `Recovery.hpp` is inappropriate and they just need to use the `nav2_core` interface, that's the real thing the factory is using.

In summary-  the `Recovery.hpp` is a set of optional convenience tools for an action server, but not all recoveries have to be action servers, nor should we require that

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp (hunk 16 above)
Comment (SteveMacenski) [reply nav2_recoveries/include/nav2_recoveries/recovery.hpp]: Sure that makes sense

File: nav2_recoveries/plugins/back_up.cpp (hunk 14 above)
Comment (SteveMacenski) [reply nav2_recoveries/plugins/back_up.cpp]: Changes have been reflected

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp (hunk 16 above)
Comment (SteveMacenski) [reply nav2_recoveries/include/nav2_recoveries/recovery.hpp]: Changes have been reflected

**Comment by bpwilcox** [comment]
> @bpwilcox the abstract interface from `nav2_core` is the _real_ recovery interface. The `Recovery.hpp` wrapper around it wraps the action server, but that's not the _real_ base interface to use, that's another level on top to reduce code duplication if that recovery is an action server, but there's nothing requiring that to be true. It could be a topic, service, action, etc.
> 
I can see that, though commenting out the `executeRecovery()` makes the interface have little value at the moment, so I would suggest we resolve that issue.

> A plugin designer could choose to use the `Recovery.hpp` wrapped interface if it makes sense, but there may be quick things that should be services, where the `Recovery.hpp` is inappropriate and they just need to use the `nav2_core` interface, that's the real thing the factory is using.
> 
I would agree, though, for example, requiring a `tf2_ros::Buffer` pointer doesn't seem like it would be a common resource for all arbitrary recoveries.

> Edit: I would not disagree that renaming `recovery.hpp`'s action server to something else would be good if that makes it more clear, which I think it would. Maybe `recovery_action_interface.hpp` or something? Let me know what you think
>
I would agree we should rename. Even in this comment I had to think about whether you were referring to the class in the `nav2_core` or the `nav2_recoveries`. Perhaps we could create a `recovery_service_interface.hpp` and whatever else makes sense.

File: nav2_recoveries/include/nav2_recoveries/recovery.hpp (hunk 17 above)
Comment (bpwilcox) [reply nav2_recoveries/include/nav2_recoveries/recovery.hpp]: I don't see where you are resetting the node and tf in the `on_cleanup` for the recovery server.

**Comment by SteveMacenski** [comment]
> though commenting out the executeRecovery() makes the interface have little value at the moment

How so? A recovery right now can just expose a service (like clear_costmap) and then have a BT node that calls that service to activate. There's no need for `executeRecovery`. That method was installed for the case of 1 recovery server action and contained within the call was the specific recovery to call from its map of recoveries. As discussed above, the action interface definition would be so large to accommodate all the possibilities, I don't know that's a great idea. Though we could make something that covers the typical basis, and then for others we expose a `createsOwnInterface()` method that tells the server whether or not its callable from its own server or the larger `nav2_recoveries` server. 

>  requiring a tf2_ros::Buffer pointer doesn't seem like it would be a common resource for all arbitrary recoveries

Totally, that's also why `nav2_core` doesn't contain a buffer member, so that if they dont want it, if they don't store it, it goes out of scope. Its similar to the costmaps in controller/planner: you make it once and give it to everyone so anyone that needs it can use it. The worst case is that all your plugins require it, and then they're all maintaining independent buffers, which would be not great. 

Renaming would be good. I've resigned that the BT is just going to be confusing since ROS and BT use the same words and in some cases both those words apply to a file, sometimes not. This we can resolve. On making a service caller, sure, I that's what I did for the BT for clearing costmaps, but that should be in another PR (and based on need, right now we have no service based recoveries).

**Comment by bpwilcox** [comment]
With the recovery interface we are aligning with the lifecycle node transitions via  `cleanup`, `activate`, etc. and passing a node and tf buffer. I'm just considering whether this could be more generic as a `LifecyclePluginInterface` that perhaps any plugin in our stack could use. 


> Totally, that's also why `nav2_core` doesn't contain a buffer member, so that if they dont want it, if they don't store it, it goes out of scope. Its similar to the costmaps in controller/planner: you make it once and give it to everyone so anyone that needs it can use it. The worst case is that all your plugins require it, and then they're all maintaining independent buffers, which would be not great.
> 
Right, I don't particularly like having to pass down a buffer that is potentially never used or stored, but I can't think of a way around it yet in creating a base class.

+1 on the renaming :)

**Comment by SteveMacenski** [comment]
@bpwilcox well the easy way of getting rid of it is loading all the plugins, checking the reference counter, and if no one wants it, we reset it. But that's potentially not what your issue is?

I would not want to have the plugin definition come with alot of baggage like what `recovery_server_interface.hpp` will have. Personally, I think its good to get people in the rhythm of using action servers more, so I'd just remove that file all together and have a clearly defined action server in each spin/backup. Its not that much boiler plate and it makes it alot more approachable so you dont have to go too far down the C++ rabbit  hole to make something. 

But that's an unpopular opinion so I think this is the best compromise. Unless you see something behind door number 3?

**Comment by bpwilcox** [comment]
> @bpwilcox well the easy way of getting rid of it is loading all the plugins, checking the reference counter, and if no one wants it, we reset it. But that's potentially not what your issue is?
> 
My issue was more about the interface requiring the tf buffer at all for the configure step

> I would not want to have the plugin definition come with alot of baggage like what `recovery_server_interface.hpp` will have. Personally, I think its good to get people in the rhythm of using action servers more, so I'd just remove that file all together and have a clearly defined action server in each spin/backup. Its not that much boiler plate and it makes it alot more approachable so you dont have to go too far down the C++ rabbit hole to make something.
> 
I'm not sure what you mean by baggage here. My point was just that having a `LifecyclePluginInterface` could homogenize the code in the `nav2_core` and make the implementation of the recovery action server base interface  (and whatever other plugin base classes) uniform with the lifecycle model, which to me, the `recovery.hpp` is essentially that except with the added tf buffer.

**Comment by SteveMacenski** [comment]
> My issue was more about the interface requiring the tf buffer at all for the configure step

I think then this is just a fundamental difference between our thought processes. I believe that any process that needs a TF buffer should get one, but exactly one. So in making it at the highest application level possible, you can make it trickle down to anyone else that needs one to ensure, no matter what random person comes along, they'll have access to one and not have to create one. When someone makes one, there's more to come and that'll bog down the memory utilization massively. Most everything will require a TF buffer at some point (to know where you are in X frame, to know where the laser is relative to the wheels, to know where a fixed waypoint is relative to the base, etc) so making sure its available is the most surefire way (in my experience) of making sure someone doesn't casually add one lower, and then another one in another place, etc. Plus its one of those fundamental building blocks we use to create applications so I don't see a reason to try to be stingy about it. 

A compromise I could see if having a `needsTF()` method and loops through before creating one to make sure someone needs it, but that seems excessive. As far as I can tell, we dont have a single plugin in the stack that _doesnt_ require the TF buffer. 

I suppose you'd have to draw out what you mean by `LifecyclePluginInterface`, we've kind of implicitly made it, if I understand you correctly, in just having the pluginlib headers defined. We could in theory abstract away some of these things with no inputs (cleanup/etc) but for states like `configure`, we're always going to want different types of plugins to be passed in different types of "stuff"

**Comment by bpwilcox** [comment]
> I think then this is just a fundamental difference between our thought processes. I believe that any process that needs a TF buffer should get one, but exactly one. So in making it at the highest application level possible, you can make it trickle down to anyone else that needs one to ensure, no matter what random person comes along, they'll have access to one and not have to create one. When someone makes one, there's more to come and that'll bog down the memory utilization massively. Most everything will require a TF buffer at some point (to know where you are in X frame, to know where the laser is relative to the wheels, to know where a fixed waypoint is relative to the base, etc) so making sure its available is the most surefire way (in my experience) of making sure someone doesn't casually add one lower, and then another one in another place, etc. Plus its one of those fundamental building blocks we use to create applications so I don't see a reason to try to be stingy about it.
> 

I don't disagree on that we shouldn't need to create more than one tf buffer. My thoughts are more towards the clean separation of plugins with like interfaces, for which in regard to recoveries, I am not convinced that all recoveries deriving from this interface would ultimately need one. Sure we could pass in one and let it go out of scope, however, I don't think that design offers the clean separation I would expect. For example, recoveryA needs tf buffer but recoveryB does not, the only common interface they have are the generic lifecycle stuff (`cleanup`, etc). My question would be then why should they derive from the same interface at all, especially if the generic lifecycle stuff could be abstracted. To your own example, if we created a "Wait"" or "Ask for Help" recovery, those probably don't need the same interface.

I think we can hash this out further maybe with future design proposals and issues, so for now I'm approving this PR.

**Comment by SteveMacenski** [comment]
Well because we have to make _something_ that they all derive from that we can create a plugin factory on. 

Sure, we could have 2 or more `recovery` plugin headers in nav2_core, but at what tangible benefit? It's another thing that someone has to figure how which to use when and why, and then we need to support multiple plugin factories in the nav2 server. I'm not saying there *isnt* a reason to do that, but I would expect if we were to add additional pluginlib definitions, it would be to add tangible capabilities.

I see where you're coming from, but I disagree that that design is more clean. It seems to be to add more complexity without a tangible benefit to the end-user. I would argue this is more clean since its more versatile. The statement that its not clean because we don't use all the inputs would be an argument against all the lifecycle nodes we have, since all the state transitions ignore the input arguments.

**Comment by SteveMacenski** [comment]
that rebase was unusually painful, lets see what CI says in the morning and merge depending
